includes/core/class-mode.php

```diff
@@ -0,0 +1,172 @@
+<?php
+/**
+ * Class file for WCPay\Core\Mode.
+ *
+ * @package WooCommerce Payments
+ */
+
+namespace WCPay\Core;
+
+use WC_Payment_Gateway_WCPay;
+use Exception;
+
+/**
+ * Controls the working mode of WooCommerce Payments.
+ *
+ * @property-read bool $live  Flag for live mode.
+ * @property-read bool $dev   Flag for dev mode.
+ * @property-read bool $test  Flag for test mode.
+ * @method        void live() Enters live mode. Should only be used for tests.
+ * @method        void dev()  Enters live mode. Should only be used for tests.
+ * @method        void test() Enters live mode. Should only be used for tests.
+ */
+class Mode {
+	/**
+	 * Holds the test mode flag.
+	 *
+	 * @var bool
+	 */
+	private $test_mode;
+
+	/**
+	 * Holds the dev mode flag.
+	 *
+	 * @var bool
+	 */
+	private $dev_mode;
+
+	/**
+	 * Holds the gateway class for settings.
+	 *
+	 * @var WC_Payment_Gateway_WCPay
+	 */
+	private $gateway;
+
+	/**
+	 * Environment types, which are used to automatically enter dev mode.
+	 *
+	 * @see wp_get_environment_type()
+	 * @see https://developer.wordpress.org/reference/functions/wp_get_environment_type/#description
+	 */
+	const DEV_MODE_ENVIRONMENTS = [
+		'development',
+		'staging',
+	];
+
+	/**
+	 * Stores the gateway for later retrieval of options.
+	 *
+	 * @param WC_Payment_Gateway_WCPay $gateway The active gateway.
+	 */
+	public function __construct( WC_Payment_Gateway_WCPay $gateway ) {
+		$this->gateway = $gateway;
+	}
+
+	/**
+	 * Initializes the working mode of WooCommerce Payments.
+	 */
+	private function init() {
+		$dev_mode = (
+			// Plugin-specific dev mode.
+			$this->is_wcpay_dev_mode_defined()
+
+			// WordPress Dev Environment.
+			|| in_array( $this->get_wp_environment_type(), self::DEV_MODE_ENVIRONMENTS, true )
+		);
+
+		/**
+		 * Allows WooCommerce to enter dev mode.
+		 *
+		 * @see https://woocommerce.com/document/payments/testing/dev-mode/
+		 * @param bool $dev_mode The pre-determined dev mode.
+		 */
+		$this->dev_mode = (bool) apply_filters( 'wcpay_dev_mode', $dev_mode );
+
+		$test_mode_setting = 'yes' === $this->gateway->get_option( 'test_mode' );
+		$test_mode         = $this->dev_mode || $test_mode_setting;
+
+		/**
+		 * Allows WooCommerce to enter test mode.
+		 *
+		 * @see https://woocommerce.com/document/payments/testing/
+		 * @param bool $test_mode The pre-determined test mode.
+		 */
+		$this->test_mode = (bool) apply_filters( 'wcpay_test_mode', $test_mode );
+	}
+
+	/**
+	 * Checks whether WCPay is in the given mode.
+	 *
+	 * @param  string $property The mode to check for. Either `live`, `test`, or `dev`.
+	 * @return bool If the propery is toggled.
+	 *
+	 * @throws Exception In case the class has not been initialized yet.
+	 */
+	public function __get( $property ) {
```

Comment:
All properties are private and this is only used inside the function and for sake of simplicity, could we create functions like is_live, is_dev, is_SOMETHING and put the logic there.

I understand this approach for an abstract classed or classes that shares common logic to other classes, but making abstractions like this could be an overkill. Yet again, maybe I am wrong, but putting a couple of functions and not relaying on magic methods makes things easier to understand and read.

Reply:
Done in https://github.com/Automattic/woocommerce-payments/pull/4986/commits/896eaa3f9aac54d505f17ac81b0d71d48a9b9534, I agree getters are not helping enough here.